feat: create inputs and outputs

rodrigoleonello opened the pull request:
Depends on:
- [ ] https://github.com/tidewise/drivers-deep_trekker/pull/3

rodrigoleonello committed:
feat: create inputs, outputs and properties.

doudou commented on deep_trekker.orogen:
```diff
@@ -1,55 +1,23 @@
-    # Data can be written using _output.write(value). _output.connected() returns
-    # if this output is connected to an input or not.
-    output_port "output", "int"
+    output_port "devices_diagnostics", "deep_trekker/Devices"
```
Shouldn't that be just the revolution data structure ?

rodrigoleonello replied:
yes, I forgot that will be one component for each device.

doudou commented on deep_trekker.orogen:
```diff
@@ -1,55 +1,23 @@
-task_context "Task" do
-    # This is the default from now on, and should not be removed. Rock will
-    # transition to a setup where all components use a configuration step.
+task_context "RevolutionInterfaceController" do
```
Maybe just `RevolutionTask` ? (in any case, always end in `Task`)

rodrigoleonello committed:
feat: query data from the devices first.

rodrigoleonello committed:
fix: change the update logic for each type of command.

Now, there are only 3 devices (revolution and each type of reel).
There is no command option for the manual reel, we will just monitor
the state.

rodrigoleonello committed:
feat: add manual reel device state to the getDevicesInfo method.

rodrigoleonello committed:
feat: add powered reel device state to the getDevicesInfo method.

rodrigoleonello committed:
feat: add input commands for camera and grabber

rodrigoleonello committed:
fix: resize data.

rodrigoleonello committed:
feat: call parser methods for each type of command.

rodrigoleonello committed:
feat: add api version as property.

and move back queryDeviceStateInfo to the init of the update (fix)

rodrigoleonello committed:
fix: modify default switch case.

rodrigoleonello committed:
feat: add test cases for each command action.

rodrigoleonello committed:
fix: update logic to receive info from vehicle after query it.

rodrigoleonello committed:
feat: update code formatting

rodrigoleonello committed:
feta: modify control command name

jhonasiv commented on test/revolution_test.rb:
```diff
@@ -0,0 +1,522 @@
+                    "578954C0A26S": {
+                        "calibrator": true,
+                        "ready": true,
+                        "distance": 15,
+                        "leak": false,
+                        "cpuTemp": 10
+                    },
+                    "63B234C0A269": {
+                        "calibrator": true,
+                        "ready": true,
+                        "distance": 15,
+                        "leak": false,
+                        "cpuTemp": 10,
+                        "battery1": {
+                            "percent": 0.8,
+                            "voltage": 1,
+                            "charging": 2
+                        },
+                        "battery2": {
+                            "percent": 0.8,
+                            "voltage": 1,
+                            "charging": 2
+                        },
+                        "motor1Diagnostics": {
+                            "current": 3,
+                            "overcurret": false,
+                            "rpm": 5.5
+                        },
+                        "motor2Diagnostics": {
+                            "current": 3,
+                            "overcurret": false,
+                            "rpm": 5.5
+                        },
+                        "speed": 5.5,
+                        "acConnected": false,
+                        "eStop": false
+                    }
+                }
+            }
+        )
```
You could turn the JSON messages into `yml` files instead of writing it all on code, I think it would be easier to interact.

jhonasiv commented on test/revolution_test.rb:
```diff
@@ -0,0 +1,522 @@
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["lights"], \
+                     json_from_expected["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["lights"]
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["lasers"], \
+                     json_from_expected["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["lasers"]
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["tilt"], \
+                     json_from_expected["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["tilt"]
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["exposure"], \
+                     json_from_expected["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["exposure"]
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["brightness"], \
+                     json_from_expected["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["brightness"]
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["focus"], \
+                     json_from_expected["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["focus"]
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["saturation"], \
+                     json_from_expected["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["saturation"]
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["sharpness"], \
+                     json_from_expected["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["sharpness"]
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["zoom"]["ratio"], \
+                     json_from_expected["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["zoom"]["ratio"]
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["zoom"]["speed"], \
+                     json_from_expected["payload"]["devices"]["57B974C0A269"] \
+            ["cameraHead"]["camera"]["zoom"]["speed"]
```
This is a nightmare to read, maybe transform part of the json in a variable? For example:

```
camera_from_sample = json_from_sample["payload"]["devices"]["57B...."]["cameraHead"]["camera"]
camera_from_expected = ...
assert_equal camera_from_sample["zoom"]["speed"], camera_from_expected["zoom"]["speed"]
```

jhonasiv commented on test/revolution_test.rb:
```diff
@@ -0,0 +1,522 @@
+        json_from_sample = JSON.parse(sample.data.to_byte_array[8..-1])
+        json_from_expected = JSON.parse(expected_raw_cmd.data.to_byte_array[8..-1])
+
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["grabber"]["openClose"], json_from_expected["payload"]["devices"] \
+            ["57B974C0A269"]["grabber"]["openClose"]
+        assert_equal json_from_sample["payload"]["devices"]["57B974C0A269"] \
+            ["grabber"]["rotate"], json_from_expected["payload"]["devices"] \
+            ["57B974C0A269"]["grabber"]["rotate"]
```
Even tests with few assertions like this one would improve from transforming the json access into variables.

doudou commented on deep_trekker.orogen:
```diff
@@ -1,55 +1,45 @@
+    # Devices command action
+    input_port "command_action", "deep_trekker/DevicesCommandAction"
+
+    # Tilt Camera head command
+    input_port "camera_head_command", "deep_trekker/TiltCameraHeadCommand"
```
Every single command and status for something that looks like a joint must be a port of type `/base/samples/Joints`. If there are additional things, create a separate port just for these things.

Commands for underwater vehicles must be a `/base/commands/LinearAngular6dCommand`

doudou replied:
You may keep the library interface (actually, I think you should). But the component interface should be updated.

doudou replied:
Body state estimate outputs (velocity, position, angular velocity, angular position) must be `base/samples/RigidBodyState` (Accelerations would be `/base/samples/RigidBodyAcceleration`)

doudou commented on tasks/RevolutionTask.cpp:
```diff
@@ -0,0 +1,222 @@
+void RevolutionTask::receiveDeviceStateInfo()
+{
+    RawPacket data_in;
+    if (_data_in.read(data_in) == RTT::NoData) {
```
`!= RTT::NewData`

rodrigoleonello committed:
fix: remove command_action input and separate the output for each device.

rodrigoleonello committed:
fix: separate joint input/output ports from each type

rodrigoleonello committed:
fix: update tests

rodrigoleonello committed:
feat: update orogen comments.

doudou commented on deep_trekker.orogen:
```diff
@@ -1,55 +1,69 @@
+    input_port "grabber_command", "base/samples/Joints"
+
+    # Vehicle setpoint in local frame and light command
+    input_port "rov2ref_command", "deep_trekker/PositionAndLightCommand"
```
Underwater vehicle commands should be /base/commands/LinearAngular6dCommand

doudou commented on deep_trekker.orogen:
```diff
@@ -1,55 +1,69 @@
+    # Rawdata out
+    output_port "data_out", "iodrivers_base/RawPacket"
 
+    # Rawdata vehicle command out (debug purpose)
+    output_port "data_command_out", "iodrivers_base/RawPacket"
+
+    # Rawdata query device info out (debug purpose)
+    output_port "data_query_out", "iodrivers_base/RawPacket"
```
We should be able to find these in `data_out` directly.

rodrigoleonello replied:
How can I do that in the test? Because always that we pass trough the update, we query data and sends data (or not) using the same output....so, what happens is that, always in the test I get just the output from the query output.

doudou replied:
You can filter the samples received with `.matching`

```
have_one_new_sample(port).matching do |s|
   json = JSON.parse(s.data.to_byte_array[8...-1])
   // some criteria on json that allows to find the right packet, e.g.
   json['type'] == "GET"
end
```

rodrigoleonello replied:
nice!

rodrigoleonello committed:
feat: separate PositionAndLightCommand into 2 inputs, in the orogen file.

doudou commented on tasks/RevolutionTask.cpp:
```diff
@@ -0,0 +1,291 @@
+{
+    mMessageParser = CommandAndStateMessageParser();
+    string get_message = mMessageParser.parseGetMessage(mAPIVersion);
+    vector<uint8_t> new_data(get_message.begin(), get_message.end());
+    RawPacket data_out;
+    data_out.time = Time::now();
+    data_out.data.resize(new_data.size());
+    data_out.data = new_data;
+    _data_out.write(data_out);
```
Once you remove the debug ports, this is `sendRawDataOutput`

rodrigoleonello committed:
fix: remove unnecessary debug outputs and update joint types.

rodrigoleonello committed:
fix: add motor overcurrent state get method and change return methods to joint, as in library

rodrigoleonello committed:
feat: update test file.

rodrigoleonello committed:
fix: remove position joint in the tilt camera head command since it is just possible to control by speed
